### ISSUE TITLE
Support session tickets on client, RFC 5077

### ISSUE THREAD

**Issue by euroelessar** [grpc/grpc#14425]
Please answer these questions before submitting your issue. 
 
### Should this be an issue in the gRPC issue tracker?
 
Create new issues for bugs and feature requests. An issue needs to be actionable. General gRPC discussions and usage questions belong to:
- [grpc.io mailing list](https://groups.google.com/forum/#!forum/grpc-io)
- [StackOverflow, with `grpc` tag](http://stackoverflow.com/questions/tagged/grpc)
 
*Please don't double post your questions in more locations; we are monitoring both channels, and the time spent de-duplicating questions is better spent answering more user questions.*
 
### What version of gRPC and what language are you using?
grpc-core 1.8.4, both python grpcio & proprietary rust binding on top of grpc-core
 
### What operating system (Linux, Windows, …) and version?
GNU/Linux
 
### What runtime / compiler are you using (e.g. python version or version of gcc)
python-2.7.14
 
### What did you do?
* Create single client pointed to server.
* Restart server on different port.
* After restart client reconnects to server.
 
### What did you expect to see?
Client should resume previously established tls session using ticket issued by server.
It happens if client uses grpc-go library.
 
### What did you see instead?
Client sends ClientHello to server with empty tls ticket extension, they have to go through full tls handshake again.

After further investigation it looks like grpc-core provides neither tickets support on client/server by default nor any means to configure it.

One of possible suggestions is to add the following interface & re-expose it from python and wire it up with boringssl session ticket support. The expectation is that user creates cache once & passes it to different channels via creation arguments.
```
#define GRPC_SSL_SESSION_CACHE_ARG "grpc.ssl_session_cache"

/** Increase reference counter for the session */
GRPCAPI void grpc_ssl_session_ref(grpc_ssl_session* session);
/** Decrease reference counter for the session */
GRPCAPI void grpc_ssl_session_unref(grpc_ssl_session* session);

/** SSL session cache for session tickets (RFC 5077).
 *  Implementation must be thread safe.
 */
typedef struct grpc_ssl_session_cache {
    /**
     * Put is called by grpc-core to store session by the given key.
     * Caller owns the session, so cache implementation should call
     * `grpc_ssl_session_ref`.
     */
    void (*put) (const char* key, grpc_ssl_session* session);
    /**
     * Get session for the given key from the cache (if any).
     * Caller owns returned session, so cache implementation should call
     * `grpc_ssl_session_ref` on session before returning one.
     */
    grpc_ssl_session* (*get) (const char* key);
} grpc_ssl_session_cache;
```

Other ideas are welcome.

**Comment by nathanielmanistaatgoogle** [comment]
@jiangtaoli2016: it looks like this is happening in [pull request 14879](https://github.com/grpc/grpc/pull/14879), which is currently blocked on a pretty big deal, and about which I hope to respond in the next day or two.

**Comment by nathanielmanistaatgoogle** [comment]
The upcoming 1.13 release will include [pull request 14879](https://github.com/grpc/grpc/pull/14879). @euroelessar, @jiangtaoli2016: is there anything to be done for this issue other than wait for the 1.13 release?

**Comment by mehrdada** [comment]
Closing this bug as 1.13 is already released. Please reopen or file another issue if there's any work item still needs to be done in this area.
